nupic/research/frameworks/continuous_learning/multihead/multihead.py

```diff
@@ -0,0 +1,205 @@
+            loss += criterion(output, target, reduction="sum").item()
+            # get the index of the max log-probability
+            pred = output.argmax(dim=1, keepdim=True)
+            total_correct += pred.eq(target.view_as(pred)).sum().item()
+
+    return {"accuracy": total_correct / len(loader.dataset),
+            "loss": loss / len(loader.dataset),
+            "total_correct": total_correct}
+
+
+def do_training(model, dataset_name, scenario, device, post_batch_callback=None):
+    """
+    Train the model.
+    :param model: pytorch model to be trained
+    :type model: torch.nn.Module
+    :param dataset_name: name of the dataset on which to train model
+    :type dataset_name: str
+    :param scenario: continuous learning setup, one of {'task', 'domain', 'class'}
+    :type scenario: str
+    :param device:
+    :type device: torch.device
+    :param post_batch_callback: function(model) to call after every batch
+    :type post_batch_callback: function
+    """
+    train_datasets, test_datasets = get_datasets(dataset_name, scenario)
+
+    train_loaders = [
+        torch.utils.data.DataLoader(train_dataset,
+                                    batch_size=TRAIN_BATCH_SIZE,
+                                    shuffle=True)
+        for train_dataset in train_datasets]
+
+    test_loaders = [
+        torch.utils.data.DataLoader(test_dataset,
+                                    batch_size=TEST_BATCH_SIZE,
+                                    shuffle=True)
+        for test_dataset in test_datasets]
+
+    n_tasks = len(train_datasets)
+
+    # optimizer for training model
+    adam = optim.Adam(model.parameters(), lr=LEARNING_RATE, betas=(0.9, 0.999))
+
+    for epoch in range(EPOCHS):
+
+        # Loop over all tasks
+        for task_num, train_loader in enumerate(train_loaders, 1):
+
+            print("Epoch {}".format(epoch))
+            train_head(model=model, loader=train_loader, optimizer=adam,
+                       criterion=F.nll_loss, device=device,
+                       active_classes=get_active_classes(task_num, scenario),
+                       post_batch_callback=post_batch_callback)
+
+            # test model
+            results = []
+            for i, _allowed_classes in enumerate([[0, 1], [2, 3], [4, 5], [6, 7],
+                                                 [8, 9]]):
+
+                allowed_classes = _allowed_classes if scenario == "task" else None
+                results = test(model=model, loader=test_loaders[i],
+                               criterion=F.nll_loss, device=device,
+                               allowed_classes=allowed_classes)
+                print("\ttask {}/{}: {}".format(i + 1, n_tasks, results))
+
+            print("Epoch {}: {}".format(epoch, results))
+
+
+def get_active_classes(task_num, scenario):
+    """
+    Return a list of label indices that are "active" during training under the given
+    scenario
+    :param task_num: the index of the task
```

Comment:
Can you elaborate a little bit on the reasoning behind the <task, class, domain> heuristic defined below? This should be somewhere in the documentation. Here is a good place to do it, the `subdataset.get_datasets` method is another alternative.

Reply:
I updated the docstring to include further explanations and a link to the continual learning benchmarks paper.